![tracker](https://us-central1-vertex-ai-mlops-369716.cloudfunctions.net/pixel-tracking?path=statmike%2Fvertex-ai-mlops%2Fdata%2Bai%2Fbq-ml%2Ffunctions%2Ftext&file=text.ipynb)
<!--- header table --->
<table align="left">
<tr>
  <td style="text-align: center">
    <a href="https://github.com/statmike/vertex-ai-mlops/blob/main/data%2Bai/bq-ml/functions/text/text.ipynb">
      <img width="32px" src="https://www.svgrepo.com/download/217753/github.svg" alt="GitHub logo">
      <br>View on<br>GitHub
    </a>
  </td>
  <td style="text-align: center">
    <a href="https://colab.research.google.com/github/statmike/vertex-ai-mlops/blob/main/data%2Bai/bq-ml/functions/text/text.ipynb">
      <img width="32px" src="https://www.gstatic.com/pantheon/images/bigquery/welcome_page/colab-logo.svg" alt="Google Colaboratory logo">
      <br>Run in<br>Colab
    </a>
  </td>
  <td style="text-align: center">
    <a href="https://console.cloud.google.com/vertex-ai/colab/import/https%3A%2F%2Fraw.githubusercontent.com%2Fstatmike%2Fvertex-ai-mlops%2Fmain%2Fdata%252Bai%2Fbq-ml%2Ffunctions%2Ftext%2Ftext.ipynb">
      <img width="32px" src="https://lh3.googleusercontent.com/JmcxdQi-qOpctIvWKgPtrzZdJJK-J3sWE1RsfjZNwshCFgE_9fULcNpuXYTilIR2hjwN" alt="Google Cloud Colab Enterprise logo">
      <br>Run in<br>Colab Enterprise
    </a>
  </td>      
  <td style="text-align: center">
    <a href="https://console.cloud.google.com/bigquery/import?url=https://github.com/statmike/vertex-ai-mlops/blob/main/data%2Bai/bq-ml/functions/text/text.ipynb">
      <img width="32px" src="https://www.gstatic.com/images/branding/gcpiconscolors/bigquery/v1/32px.svg" alt="BigQuery logo">
      <br>Open in<br>BigQuery Studio
    </a>
  </td>
  <td style="text-align: center">
    <a href="https://console.cloud.google.com/vertex-ai/workbench/deploy-notebook?download_url=https://raw.githubusercontent.com/statmike/vertex-ai-mlops/main/data%2Bai/bq-ml/functions/text/text.ipynb">
      <img width="32px" src="https://www.gstatic.com/images/branding/gcpiconscolors/vertexai/v1/32px.svg" alt="Vertex AI logo">
      <br>Open in<br>Vertex AI Workbench
    </a>
  </td>
</tr>
<tr>
  <td colspan="5" style="text-align: right">
    <b>Share This On: </b> 
    <a href="https://www.linkedin.com/sharing/share-offsite/?url=https://github.com/statmike/vertex-ai-mlops/blob/main/data%252Bai/bq-ml/functions/text/text.ipynb"><img src="https://upload.wikimedia.org/wikipedia/commons/8/81/LinkedIn_icon.svg" alt="Linkedin Logo" width="20px"></a> 
    <a href="https://reddit.com/submit?url=https://github.com/statmike/vertex-ai-mlops/blob/main/data%252Bai/bq-ml/functions/text/text.ipynb"><img src="https://redditinc.com/hubfs/Reddit%20Inc/Brand/Reddit_Logo.png" alt="Reddit Logo" width="20px"></a> 
    <a href="https://bsky.app/intent/compose?text=https://github.com/statmike/vertex-ai-mlops/blob/main/data%252Bai/bq-ml/functions/text/text.ipynb"><img src="https://upload.wikimedia.org/wikipedia/commons/7/7a/Bluesky_Logo.svg" alt="BlueSky Logo" width="20px"></a> 
    <a href="https://twitter.com/intent/tweet?url=https://github.com/statmike/vertex-ai-mlops/blob/main/data%252Bai/bq-ml/functions/text/text.ipynb"><img src="https://upload.wikimedia.org/wikipedia/commons/5/5a/X_icon_2.svg" alt="X (Twitter) Logo" width="20px"></a> 
  </td>
</tr>
<tr>
  <td colspan="5" style="text-align: right">
    <b>Connect With Author On: </b> 
    <a href="https://www.linkedin.com/in/statmike"><img src="https://upload.wikimedia.org/wikipedia/commons/8/81/LinkedIn_icon.svg" alt="Linkedin Logo" width="20px"></a>
    <a href="https://www.github.com/statmike"><img src="https://www.svgrepo.com/download/217753/github.svg" alt="GitHub Logo" width="20px"></a> 
    <a href="https://www.youtube.com/@statmike-channel"><img src="https://upload.wikimedia.org/wikipedia/commons/f/fd/YouTube_full-color_icon_%282024%29.svg" alt="YouTube Logo" width="20px"></a>
    <a href="https://bsky.app/profile/statmike.bsky.social"><img src="https://upload.wikimedia.org/wikipedia/commons/7/7a/Bluesky_Logo.svg" alt="BlueSky Logo" width="20px"></a> 
    <a href="https://x.com/statmike"><img src="https://upload.wikimedia.org/wikipedia/commons/5/5a/X_icon_2.svg" alt="X (Twitter) Logo" width="20px"></a>
  </td>
</tr>
<tr>
  <td colspan="5" style="text-align: right">
    <a href="https://raw.githubusercontent.com/statmike/vertex-ai-mlops/main/data%2Bai/bq-ml/functions/text/text.ipynb"><img src="https://www.svgrepo.com/download/5445/download-button.svg" alt="Download icon" width="20px"></a> <a href="https://raw.githubusercontent.com/statmike/vertex-ai-mlops/main/data%2Bai/bq-ml/functions/text/text.ipynb">Download Notebook</a> <i>(right-click and "Save As")</i>
  </td>
</tr>
</table>

# Text — BigQuery ML Model-Free Functions

Three functions that turn tokenized text (`ARRAY<STRING>`) into ML features: `ML.NGRAMS` (scalar — builds n-grams from tokens), `ML.TF_IDF`, and `ML.BAG_OF_WORDS` (both **analytic** — require `OVER()`, build a dictionary across the whole document window). `ML.TF_IDF` and `ML.BAG_OF_WORDS` had **no example anywhere in this repo** before this notebook.

> **GOTCHA:** GoogleSQL also has same-named non-`ML.*` [text-analysis functions](https://cloud.google.com/bigquery/docs/reference/standard-sql/text-analysis-functions) (`TF_IDF`, `BAG_OF_WORDS`) with **different** semantics (term-string dictionary index, frequency-ordered). The `ML.*` versions documented here use integer dictionary indices, order the dictionary alphabetically, and reserve index `0` for the unknown term. Always call them as `ML.TF_IDF` / `ML.BAG_OF_WORDS`.

**When to use these:**
- `ML.NGRAMS` — capture local token order (bigrams, trigrams) that single tokens lose, before bag-of-words or TF-IDF.
- `ML.BAG_OF_WORDS` — raw per-document term counts, a quick baseline text feature.
- `ML.TF_IDF` — weight terms by importance (frequent-in-doc but rare-across-corpus), better than raw counts for sparse text features feeding a linear model.

**Data:** [`bigquery-public-data.thelook_ecommerce.products`](https://console.cloud.google.com/marketplace/product/bigquery-public-datasets)`.name` — short, real product-name strings, already used in [`workflows/embeddings_classification/`](../../workflows/embeddings_classification/).

**References:** [Full reference](../../RESOURCES.md) | [`ML.NGRAMS` docs](https://cloud.google.com/bigquery/docs/reference/standard-sql/bigqueryml-syntax-ngrams) | [`ML.TF_IDF` docs](https://cloud.google.com/bigquery/docs/reference/standard-sql/bigqueryml-syntax-tf-idf) | [`ML.BAG_OF_WORDS` docs](https://cloud.google.com/bigquery/docs/reference/standard-sql/bigqueryml-syntax-bag-of-words) | [Setup guide](../../setup/)

---
## Setup

Set your project and location, authenticate, and create a shared dataset. No connection needed.

In [1]:
PROJECT_ID = 'statmike-mlops-349915'  # <-- Replace with your project ID
LOCATION = 'US'  # BigQuery dataset location
DATASET_ID = 'bq_ml'  # Shared dataset across all bq-ml notebooks

### Environment

> **Already set up the project environment?** The cell below is a no-op — packages are already in your kernel. See the [Setup Reference](../../setup/) for details.
>
> **Running standalone** (Colab, Colab Enterprise, Vertex AI Workbench)? The cell below installs required packages into your current kernel.

In [2]:
import subprocess, sys, shutil

def install(*packages):
    """Install packages using uv (fast) with pip fallback."""
    uv = shutil.which('uv')
    if uv:
        subprocess.check_call([uv, 'pip', 'install', '-q', '--python', sys.executable, *packages])
    else:
        subprocess.check_call([sys.executable, '-m', 'pip', 'install', '-q', '--upgrade', *packages])

install('google-cloud-bigquery', 'db-dtypes', 'bigquery-magics', 'bigframes')

In [3]:
# Authenticate (Colab only — other environments use Application Default Credentials)
try:
    from google.colab import auth
    auth.authenticate_user()
except ImportError:
    pass  # Not in Colab — ADC is used automatically

In [4]:
from google.cloud import bigquery
import pandas as pd

client = bigquery.Client(project=PROJECT_ID)
pd.set_option('display.max_colwidth', None)

# Create the shared dataset (idempotent)
dataset_ref = bigquery.DatasetReference(PROJECT_ID, DATASET_ID)
dataset = bigquery.Dataset(dataset_ref)
dataset.location = LOCATION
client.create_dataset(dataset, exists_ok=True)
print(f'Dataset {PROJECT_ID}.{DATASET_ID} ready')

# Register %%bigquery cell magic (auto-loaded in Colab, needed elsewhere)
%load_ext bigquery_magics

Dataset statmike-mlops-349915.bq_ml ready


---
## Step 1 — Tokenize first: all three functions operate on `ARRAY<STRING>`

None of these functions take raw text directly — split it into tokens first (here, a simple lowercase + space split; a real pipeline might use GoogleSQL's `TEXT_ANALYZE`). This step exists because `ML.NGRAMS`/`ML.TF_IDF`/`ML.BAG_OF_WORDS` all work by counting or concatenating **discrete units** — there's no way to count "how many times does a word appear" or "merge adjacent words" on a single undivided string; the string has to be broken into a list of individual tokens first.

In [5]:
query = """
SELECT name, SPLIT(LOWER(name), ' ') AS tokens
FROM `bigquery-public-data.thelook_ecommerce.products`
LIMIT 3
"""
client.query(query).to_dataframe()

,name,tokens
0,Low Profile Dyed Cotton Twill Cap - Navy W39S55D,"[low, profile, dyed, cotton, twill, cap, -, navy, w39s55d]"
1,Low Profile Dyed Cotton Twill Cap - Putty W39S55D,"[low, profile, dyed, cotton, twill, cap, -, putty, w39s55d]"
2,Enzyme Regular Solid Army Caps-Black W35S45D,"[enzyme, regular, solid, army, caps-black, w35s45d]"


---
## Step 2 — `ML.NGRAMS`: scalar, builds n-grams within a size range

`range=[2,3]` returns every bigram and trigram, each `separator`-joined (default: a single space).

> **GOTCHA (verified live, corrects RESOURCES.md):** the docs describe `range` as accepting "a single integer `x`" to mean `[x, x]` — **this is not actually true.** Passing a bare `INT64` (e.g. `ML.NGRAMS(tokens, 2)`) fails outright: `"No matching signature for function ML.NGRAMS... Unable to coerce type INT64 to expected type ARRAY<INT64>"`. `range` must always be an `ARRAY<INT64>`, even for a single size — use `[2, 2]`, not `2`.

In [6]:
query = """
SELECT
  ML.NGRAMS(tokens, [2, 3]) AS default_separator,
  ML.NGRAMS(tokens, [2, 3], '_') AS underscore_separator,
  ML.NGRAMS(tokens, [2, 2]) AS bigrams_only
FROM (SELECT SPLIT(LOWER('Low Profile Dyed Cotton Cap'), ' ') AS tokens)
"""
client.query(query).to_dataframe()

,default_separator,underscore_separator,bigrams_only
0,"[low profile, low profile dyed, profile dyed, profile dyed cotton, dyed cotton, dyed cotton cap, cotton cap]","[low_profile, low_profile_dyed, profile_dyed, profile_dyed_cotton, dyed_cotton, dyed_cotton_cap, cotton_cap]","[low profile, profile dyed, dyed cotton, cotton cap]"


---
## Step 3 — `ML.BAG_OF_WORDS` and `ML.TF_IDF` on real product-name tokens

Both are analytic — the dictionary is fit across every document in the `OVER()` window, not per-row. `BAG_OF_WORDS` returns raw counts; `TF_IDF` weights them by cross-document rarity.

In [7]:
query = """
WITH docs AS (
  SELECT id, SPLIT(LOWER(REGEXP_REPLACE(name, r'[^a-zA-Z0-9 ]', ' ')), ' ') AS tokens
  FROM `bigquery-public-data.thelook_ecommerce.products`
  WHERE category = 'Accessories'
  LIMIT 500
)
SELECT id, tokens,
  ML.BAG_OF_WORDS(tokens) OVER() AS bow_default,
  ML.TF_IDF(tokens) OVER() AS tfidf_default
FROM docs
LIMIT 5
"""
client.query(query).to_dataframe()

,id,tokens,bow_default,tfidf_default
0,29033,"[hde, pattern, suspenders]","[{'index': 60, 'value': 1.0}, {'index': 97, 'value': 1.0}, {'index': 140, 'value': 1.0}]","[{'index': 60, 'value': 1.4782590666864701}, {'index': 97, 'value': 1.4275336181786233}, {'index': 140, 'value': 1.4275336181786233}]"
1,14184,"[hello, kitty, faces, black, seatbelt, belt]","[{'index': 0, 'value': 4.0}, {'index': 13, 'value': 1.0}, {'index': 15, 'value': 1.0}]","[{'index': 0, 'value': 2.558917120418325}, {'index': 13, 'value': 0.4299745353570289}, {'index': 15, 'value': 0.3880301066971682}]"
2,28896,"[s1, it, smooth, medium, eyeglass, case]","[{'index': 0, 'value': 3.0}, {'index': 26, 'value': 1.0}, {'index': 42, 'value': 1.0}, {'index': 126, 'value': 1.0}]","[{'index': 0, 'value': 1.919187840313744}, {'index': 26, 'value': 0.6725372398690506}, {'index': 42, 'value': 0.7391295333432351}, {'index': 126, 'value': 0.6918399843164409}]"
3,28348,"[alexander, julian, men, s, boxed, tie, gift, set]","[{'index': 0, 'value': 4.0}, {'index': 52, 'value': 1.0}, {'index': 81, 'value': 1.0}, {'index': 112, 'value': 1.0}, {'index': 117, 'value': 1.0}]","[{'index': 0, 'value': 1.919187840313744}, {'index': 52, 'value': 0.5353251068169838}, {'index': 81, 'value': 0.20958626743040848}, {'index': 112, 'value': 0.17198566815064437}, {'index': 117, 'value': 0.5188799882373307}]"
4,13605,"[women, s, italian, cashmere, lined, leather, gloves, by, fratelli, orsini, everyday, , 17, colors, ]","[{'index': 0, 'value': 1.0}, {'index': 1, 'value': 2.0}, {'index': 23, 'value': 1.0}, {'index': 27, 'value': 1.0}, {'index': 32, 'value': 1.0}, {'index': 41, 'value': 1.0}, {'index': 50, 'value': 1.0}, {'index': 54, 'value': 1.0}, {'index': 69, 'value': 1.0}, {'index': 77, 'value': 1.0}, {'index': 79, 'value': 1.0}, {'index': 93, 'value': 1.0}, {'index': 112, 'value': 1.0}, {'index': 164, 'value': 1.0}]","[{'index': 0, 'value': 0.25589171204183253}, {'index': 1, 'value': 0.21631495512200477}, {'index': 23, 'value': 0.22397356562422122}, {'index': 27, 'value': 0.25022786501676714}, {'index': 32, 'value': 0.21110834230162154}, {'index': 41, 'value': 0.295651813337294}, {'index': 50, 'value': 0.28550672363572466}, {'index': 54, 'value': 0.1406515953019381}, {'index': 69, 'value': 0.28550672363572466}, {'index': 77, 'value': 0.16906061361602484}, {'index': 79, 'value': 0.23574111330410175}, {'index': 93, 'value': 0.28550672363572466}, {'index': 112, 'value': 0.09172568968034366}, {'index': 164, 'value': 0.13426912297696103}]"


**The concrete difference between the two, isolated:** 4 tiny documents, each containing `'common_word'` (in all 4) plus one other word (`'apple'`/`'banana'`/`'rare_word'`/`'rare_word'` — appearing in 1-2 docs). Lower the threshold so nothing gets dropped to bucket 0 (`top_k=100, frequency_threshold=1`), isolating the weighting behavior itself:

In [8]:
query = """
WITH docs AS (
  SELECT * FROM UNNEST([
    STRUCT(1 AS doc_id, ['common_word', 'apple'] AS tokens),
    STRUCT(2 AS doc_id, ['common_word', 'banana'] AS tokens),
    STRUCT(3 AS doc_id, ['common_word', 'rare_word'] AS tokens),
    STRUCT(4 AS doc_id, ['common_word', 'rare_word'] AS tokens)
  ])
)
SELECT doc_id, tokens,
  ML.BAG_OF_WORDS(tokens, 100, 1) OVER() AS bow_counts,
  ML.TF_IDF(tokens, 100, 1) OVER() AS tfidf_weights
FROM docs
ORDER BY doc_id
"""
client.query(query).to_dataframe()

,doc_id,tokens,bow_counts,tfidf_weights
0,1,"[common_word, apple]","[{'index': 1, 'value': 1.0}, {'index': 3, 'value': 1.0}]","[{'index': 1, 'value': 0.5493061443340549}, {'index': 3, 'value': 0.29389333245105953}]"
1,2,"[common_word, banana]","[{'index': 2, 'value': 1.0}, {'index': 3, 'value': 1.0}]","[{'index': 2, 'value': 0.5493061443340549}, {'index': 3, 'value': 0.29389333245105953}]"
2,3,"[common_word, rare_word]","[{'index': 3, 'value': 1.0}, {'index': 4, 'value': 1.0}]","[{'index': 3, 'value': 0.29389333245105953}, {'index': 4, 'value': 0.4236489301936017}]"
3,4,"[common_word, rare_word]","[{'index': 3, 'value': 1.0}, {'index': 4, 'value': 1.0}]","[{'index': 3, 'value': 0.29389333245105953}, {'index': 4, 'value': 0.4236489301936017}]"


**Verified:** `bow_counts` gives `'common_word'` (appears in all 4 docs) the exact same weight (`1.0`) as `'apple'`/`'banana'`/`'rare_word'` (rarer) — raw counts don't distinguish importance. `tfidf_weights` gives `'common_word'` a **lower** weight (`0.294`) than `'apple'`/`'banana'` (`0.549`, appearing in only 1 doc each) or `'rare_word'` (`0.424`, appearing in 2 docs) — TF-IDF genuinely down-weights terms that show up everywhere and up-weights rarer, more distinguishing ones.

---
## Step 4 — MAJOR GOTCHA (verified live, first time tested anywhere in this repo)

`ML.TF_IDF`/`ML.BAG_OF_WORDS` share the exact same `top_k=32000`/`frequency_threshold=5` defaults as the encoders in [`functions/encoding/`](../encoding/) — and the **same practical consequence, verified for both functions, not just asserted by analogy**. Six tiny "documents," each containing `'common_word'` (appears in all 6, ≥5) plus one other word that appears in only 1-2 documents (below the default threshold of 5):

In [9]:
query = """
WITH docs AS (
  SELECT * FROM UNNEST([
    STRUCT(1 AS doc_id, ['common_word', 'apple'] AS tokens),
    STRUCT(2 AS doc_id, ['common_word', 'banana'] AS tokens),
    STRUCT(3 AS doc_id, ['common_word', 'rare_word'] AS tokens),
    STRUCT(4 AS doc_id, ['common_word', 'cherry'] AS tokens),
    STRUCT(5 AS doc_id, ['common_word', 'date'] AS tokens),
    STRUCT(6 AS doc_id, ['common_word', 'rare_word'] AS tokens)
  ])
)
SELECT doc_id, tokens,
  ML.BAG_OF_WORDS(tokens) OVER() AS bow_default_freq5,
  ML.BAG_OF_WORDS(tokens, 100, 1) OVER() AS bow_freqthresh1,
  ML.TF_IDF(tokens) OVER() AS tfidf_default_freq5,
  ML.TF_IDF(tokens, 100, 1) OVER() AS tfidf_freqthresh1
FROM docs
ORDER BY doc_id
"""
client.query(query).to_dataframe()

,doc_id,tokens,bow_default_freq5,bow_freqthresh1,tfidf_default_freq5,tfidf_freqthresh1
0,1,"[common_word, apple]","[{'index': 0, 'value': 1.0}, {'index': 1, 'value': 1.0}]","[{'index': 1, 'value': 1.0}, {'index': 4, 'value': 1.0}]","[{'index': 0, 'value': 0.30951960420311175}, {'index': 1, 'value': 0.30951960420311175}]","[{'index': 1, 'value': 0.6931471805599453}, {'index': 4, 'value': 0.30951960420311175}]"
1,2,"[common_word, banana]","[{'index': 0, 'value': 1.0}, {'index': 1, 'value': 1.0}]","[{'index': 2, 'value': 1.0}, {'index': 4, 'value': 1.0}]","[{'index': 0, 'value': 0.30951960420311175}, {'index': 1, 'value': 0.30951960420311175}]","[{'index': 2, 'value': 0.6931471805599453}, {'index': 4, 'value': 0.30951960420311175}]"
2,3,"[common_word, rare_word]","[{'index': 0, 'value': 1.0}, {'index': 1, 'value': 1.0}]","[{'index': 4, 'value': 1.0}, {'index': 6, 'value': 1.0}]","[{'index': 0, 'value': 0.30951960420311175}, {'index': 1, 'value': 0.30951960420311175}]","[{'index': 4, 'value': 0.30951960420311175}, {'index': 6, 'value': 0.5493061443340549}]"
3,4,"[common_word, cherry]","[{'index': 0, 'value': 1.0}, {'index': 1, 'value': 1.0}]","[{'index': 3, 'value': 1.0}, {'index': 4, 'value': 1.0}]","[{'index': 0, 'value': 0.30951960420311175}, {'index': 1, 'value': 0.30951960420311175}]","[{'index': 3, 'value': 0.6931471805599453}, {'index': 4, 'value': 0.30951960420311175}]"
4,5,"[common_word, date]","[{'index': 0, 'value': 1.0}, {'index': 1, 'value': 1.0}]","[{'index': 4, 'value': 1.0}, {'index': 5, 'value': 1.0}]","[{'index': 0, 'value': 0.30951960420311175}, {'index': 1, 'value': 0.30951960420311175}]","[{'index': 4, 'value': 0.30951960420311175}, {'index': 5, 'value': 0.6931471805599453}]"
5,6,"[common_word, rare_word]","[{'index': 0, 'value': 1.0}, {'index': 1, 'value': 1.0}]","[{'index': 4, 'value': 1.0}, {'index': 6, 'value': 1.0}]","[{'index': 0, 'value': 0.30951960420311175}, {'index': 1, 'value': 0.30951960420311175}]","[{'index': 4, 'value': 0.30951960420311175}, {'index': 6, 'value': 0.5493061443340549}]"


**Verified for both functions:** under the default `frequency_threshold=5`, **every** word except `'common_word'` collapses to index `0` in both `ML.BAG_OF_WORDS` AND `ML.TF_IDF` — even `'rare_word'`, which appears in **2 different documents**, is indistinguishable from `'apple'` (1 occurrence) or a truly unseen term. With `frequency_threshold=1`, all 6 distinct words get their own index in both. Same practical consequence as the encoder gotcha in `functions/encoding/`: on a small corpus, almost everything except the most common terms silently disappears under current defaults.

---
## Step 5 — Embedded in a real `CREATE MODEL TRANSFORM`: text classification

`ML.BAG_OF_WORDS` features derived directly from the product name, feeding a `LOGISTIC_REG` that distinguishes two categories.

In [10]:
query = """
CREATE OR REPLACE MODEL `{project}.{dataset}.text_downstream_logistic_regression`
TRANSFORM(
  category,
  ML.BAG_OF_WORDS(SPLIT(LOWER(REGEXP_REPLACE(name, r'[^a-zA-Z0-9 ]', ' ')), ' '), 200, 3) OVER() AS name_bow
)
OPTIONS(model_type = 'LOGISTIC_REG', input_label_cols = ['category']) AS
SELECT category, name
FROM `bigquery-public-data.thelook_ecommerce.products`
WHERE category IN ('Accessories', 'Jeans')
""".format(project=PROJECT_ID, dataset=DATASET_ID)
client.query(query).result()
print('Model text_downstream_logistic_regression created')

query = f"SELECT * FROM ML.EVALUATE(MODEL `{PROJECT_ID}.{DATASET_ID}.text_downstream_logistic_regression`)"
client.query(query).to_dataframe()

Model text_downstream_logistic_regression created


,precision,recall,accuracy,f1_score,log_loss,roc_auc
0,0.91762,0.992574,0.947083,0.953627,0.547669,0.996865


~0.95 accuracy — bag-of-words features from the product name alone strongly separate `Accessories` from `Jeans`, with no manual feature engineering beyond tokenizing.

---
## Examples — `%%bigquery` Magics

The same operations using IPython magic commands — write SQL directly in cells without Python string wrapping.

In [11]:
%%bigquery --project {PROJECT_ID}

SELECT tokens, ML.NGRAMS(tokens, [2, 3]) AS bigrams_trigrams
FROM (SELECT SPLIT(LOWER('Low Profile Dyed Cotton Cap'), ' ') AS tokens)

Query is running:   0%|          |

Downloading:   0%|          |

,tokens,bigrams_trigrams
0,"[low, profile, dyed, cotton, cap]","[low profile, low profile dyed, profile dyed, profile dyed cotton, dyed cotton, dyed cotton cap, cotton cap]"


---
## Examples — BigFrames

There is **no** direct BigFrames equivalent for any of these three — use the SQL `TRANSFORM` path, or BigFrames' own SQL passthrough (`bpd.read_gbq`) as shown below.

In [12]:
import bigframes.pandas as bpd

bpd.close_session()  # Reset session to apply project/location settings
bpd.options.bigquery.project = PROJECT_ID
bpd.options.bigquery.location = LOCATION

bf_query = """
SELECT tokens, ML.NGRAMS(tokens, [2, 3]) AS bigrams_trigrams
FROM (SELECT SPLIT(LOWER('Low Profile Dyed Cotton Cap'), ' ') AS tokens)
"""
bpd.read_gbq(bf_query).peek()

,tokens,bigrams_trigrams
0,['low' 'profile' 'dyed' 'cotton' 'cap'],['low profile' 'low profile dyed' 'profile dyed' 'profile dyed cotton' 'dyed cotton' 'dyed cotton cap' 'cotton cap']


---
## Cleanup

Drop the model created in Step 5. The shared dataset is left for other notebooks.

In [13]:
client.query(f"DROP MODEL IF EXISTS `{PROJECT_ID}.{DATASET_ID}.text_downstream_logistic_regression`").result()
print('Dropped text_downstream_logistic_regression')

Dropped text_downstream_logistic_regression


### Remove all shared resources

When finished with **all** notebooks, uncomment and run the cell below to delete the shared dataset and everything in it.

In [14]:
# client.delete_dataset(
#     f'{PROJECT_ID}.{DATASET_ID}',
#     delete_contents=True,
#     not_found_ok=True
# )
# print(f'Dataset {PROJECT_ID}.{DATASET_ID} deleted')